File: backend/src/main/java/com/tunisia/financial/service/UserService.java
package com.tunisia.financial.service;

import com.tunisia.financial.dto.*;
import com.tunisia.financial.entity.User;
import com.tunisia.financial.enumerations.UserRole;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Service interface for user management operations
 * Defines business logic for user registration, authentication, and account management
 */
public interface UserService {
    
    /**
     * Register a new user
     * Creates user account
     * 
     * @param request User registration data
     * @return Created user response
     * @throws com.tunisia.financial.exception.UserAlreadyExistsException if email already exists
     */
    UserResponse registerUser(UserRegistrationRequest request);
    
    /**
     * Authenticate user with credentials
     * 
     * @param request Login credentials
     * @return Login response with JWT tokens
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     * @throws com.tunisia.financial.exception.InvalidPasswordException if password is incorrect
     * @throws com.tunisia.financial.exception.AccountLockedException if account is locked
     */
    LoginResponse login(LoginRequest request);
    
    /**
     * Get user by ID
     * 
     * @param userId User's unique identifier
     * @return User response
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     */
    UserResponse getUserById(UUID userId);
    
    /**
     * Get user by email
     * 
     * @param email User's email address
     * @return User response
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     */
    UserResponse getUserByEmail(String email);
    
    /**
     * Update user information
     * 
     * @param userId User's ID
     * @param request Update data
     * @return Updated user response
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     */
    UserResponse updateUser(UUID userId, UserUpdateRequest request);
    
    /**
     * Change user password
     * 
     * @param userId User's ID
     * @param request Password change data
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     * @throws com.tunisia.financial.exception.InvalidPasswordException if current password is incorrect
     */
    void changePassword(UUID userId, PasswordChangeRequest request);
    
    /**
     * Lock user account
     * 
     * @param userId User's ID
     * @param durationMinutes Lock duration in minutes
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     */
    void lockAccount(UUID userId, int durationMinutes);
    
    /**
     * Unlock user account
     * 
     * @param userId User's ID
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     */
    void unlockAccount(UUID userId);
    
    /**
     * Delete user account
     * 
     * @param userId User's ID
     * @throws com.tunisia.financial.exception.UserNotFoundException if user not found
     */
    void deleteUser(UUID userId);
    
    /**
     * Get all users with pagination
     * 
     * @param pageable Pagination parameters
     * @return Page of users
     */
    Page<UserResponse> getAllUsers(Pageable pageable);
    
    /**
     * Get users by role
     * 
     * @param role User role
     * @param pageable Pagination parameters
     * @return Page of users with specified role
     */
    Page<UserResponse> getUsersByRole(UserRole role, Pageable pageable);
    
    /**
     * Search users by email or name
     * 
     * @param searchTerm Search term
     * @param pageable Pagination parameters
     * @return Page of matching users
     */
    Page<UserResponse> searchUsers(String searchTerm, Pageable pageable);
    
    /**
     * Get user statistics
     * 
     * @return User statistics (counts by role, verified, locked, etc.)
     */
    UserStatistics getUserStatistics();
    
    /**
     * Convert User entity to UserResponse DTO
     * 
     * @param user User entity
     * @return User response DTO
     */
    UserResponse toUserResponse(User user);
}

File: frontend/src/services/userService.ts
import api from './api';
import {
  UserResponse,
  UserRegistrationRequest,
  UserUpdateRequest,
  PasswordChangeRequest,
  UserStatistics,
  UserRole
} from '../types/user.types';
import { PaginatedResponse, SearchParams } from '../types/common.types';

const API_BASE = '/users';

export const userService = {
  // Get user by ID
  getUserById: async (userId: string): Promise<UserResponse> => {
    const response = await api.get(`${API_BASE}/${userId}`);
    return response.data;
  },

  // Update user
  updateUser: async (userId: string, data: UserUpdateRequest): Promise<UserResponse> => {
    const response = await api.put(`${API_BASE}/${userId}`, data);
    return response.data;
  },

  // Lock user account (Admin only)
  lockUser: async (userId: string): Promise<void> => {
    await api.post(`${API_BASE}/${userId}/lock`);
  },

  // Unlock user account (Admin only)
  unlockUser: async (userId: string): Promise<void> => {
    await api.post(`${API_BASE}/${userId}/unlock`);
  },

  // Delete user (Admin only)
  deleteUser: async (userId: string): Promise<void> => {
    await api.delete(`${API_BASE}/${userId}`);
  },

  // Get all users (Admin/Auditor)
  getAllUsers: async (params?: SearchParams): Promise<PaginatedResponse<UserResponse>> => {
    const response = await api.get(`${API_BASE}`, { params });
    return response.data;
  },

  // Search users
  searchUsers: async (query: string, params?: SearchParams): Promise<PaginatedResponse<UserResponse>> => {
    const response = await api.get(`${API_BASE}/search`, {
      params: { query, ...params }
    });
    return response.data;
  },

  // Get users by role
  getUsersByRole: async (role: UserRole, params?: SearchParams): Promise<PaginatedResponse<UserResponse>> => {
    const response = await api.get(`${API_BASE}/by-role/${role}`, { params });
    return response.data;
  },

  // Get user statistics (Admin only)
  getUserStatistics: async (): Promise<UserStatistics> => {
    const response = await api.get(`${API_BASE}/statistics`);
    return response.data;
  },

  // Register new user (Admin only)
  registerUser: async (data: UserRegistrationRequest): Promise<UserResponse> => {
    const response = await api.post('/api/v1/auth/register', data);
    return response.data;
  },

  // Change password
  changePassword: async (data: PasswordChangeRequest): Promise<void> => {
    await api.post('/api/v1/auth/change-password', data);
  }
};
